PreFreeSurfer/PreFreeSurferPipeline.sh

```diff
@@ -452,253 +531,327 @@ log_Msg "POSIXLY_CORRECT="${POSIXLY_CORRECT}
+          OutputTXwImageSTRING="${OutputTXwImageSTRING}${TXwFolder}/${TXwImage}${i}_gdc "
+          i=$(($i+1))
+        done
+
+      else
+        log_Msg "NOT PERFORMING GRADIENT DISTORTION CORRECTION"
+
+        i=1
+        for Image in $TXwInputImages ; do
+          ${RUN} ${FSLDIR}/bin/fslreorient2std $Image ${TXwFolder}/${TXwImage}${i}_gdc
+          OutputTXwImageSTRING="${OutputTXwImageSTRING}${TXwFolder}/${TXwImage}${i}_gdc "
+          i=$(($i+1))
+        done
+
+      fi
+
+      # Average Like (Same Modality) Scans
+
+      if [ `echo $TXwInputImages | wc -w` -gt 1 ] ; then
+        log_Msg "Averaging ${TXw} Images"
+        log_Msg "mkdir -p ${TXwFolder}/Average${TXw}Images"
+        mkdir -p ${TXwFolder}/Average${TXw}Images
+        log_Msg "PERFORMING SIMPLE AVERAGING"
+        ${RUN} ${HCPPIPEDIR_PreFS}/AnatomicalAverage.sh -o ${TXwFolder}/${TXwImage} -s ${TXwTemplate} -m ${TemplateMask} -n -w ${TXwFolder}/Average${TXw}Images --noclean -v -b $BrainSize $OutputTXwImageSTRING
+      else
+        log_Msg "Not Averaging ${TXw} Images"
+        log_Msg "ONLY ONE AVERAGE FOUND: COPYING"
+        ${RUN} ${FSLDIR}/bin/imcp ${TXwFolder}/${TXwImage}1_gdc ${TXwFolder}/${TXwImage}
+      fi
+
+      # ACPC align T1w or T2w image to specified MNI Template to create native volume space
+      log_Msg "Aligning ${TXw} image to ${TXwTemplate} to create native volume space"
+      log_Msg "mkdir -p ${TXwFolder}/ACPCAlignment"
+      mkdir -p ${TXwFolder}/ACPCAlignment
+      ${RUN} ${HCPPIPEDIR_PreFS}/ACPCAlignment.sh \
+        --workingdir=${TXwFolder}/ACPCAlignment \
+        --in=${TXwFolder}/${TXwImage} \
+        --ref=${TXwTemplate} \
+        --out=${TXwFolder}/${TXwImage}_acpc \
+        --omat=${TXwFolder}/xfms/acpc.mat \
+        --brainsize=${BrainSize}
+
+      # Brain Extraction(FNIRT-based Masking)
+      log_Msg "Performing Brain Extraction using FNIRT-based Masking"
+      log_Msg "mkdir -p ${TXwFolder}/BrainExtraction_FNIRTbased"
+      mkdir -p ${TXwFolder}/BrainExtraction_FNIRTbased
+      ${RUN} ${HCPPIPEDIR_PreFS}/BrainExtraction_FNIRTbased.sh \
+        --workingdir=${TXwFolder}/BrainExtraction_FNIRTbased \
+        --in=${TXwFolder}/${TXwImage}_acpc \
+        --ref=${TXwTemplate} \
+        --refmask=${TemplateMask} \
+        --ref2mm=${TXwTemplate2mm} \
+        --ref2mmmask=${Template2mmMask} \
+        --outbrain=${TXwFolder}/${TXwImage}_acpc_brain \
+        --outbrainmask=${TXwFolder}/${TXwImage}_acpc_brain_mask \
+      --fnirtconfig=${FNIRTConfig}
+
+  done
+
+  # End of looping over modalities (T1w and T2w)
+
+  # ------------------------------------------------------------------------------
+  #  T2w to T1w Registration and Optional Readout Distortion Correction
+  # ------------------------------------------------------------------------------
+
+  if [ "${T2wInputImages}" = "NONE" ] ; then
+    T2wImagePath="NONE"
+  else
+    T2wImagePath="${T2wFolder}/${T2wImage}_acpc"
```

Comment:
This variable is not a "path", but rather a NIFTI image, so I'm not a fan of the variable name here.  Also, turning this into a variable messes up the symmetry of the call to `T2wToT1wDistortionCorrectAndReg.sh` in a way that makes it harder to interpret.

Reply:
I have now removed the additional code and brought the calls to `T2wToT1wDistortionCorrectAndReg.sh` and `T2wToT1wReg.sh` to their previous state. In them I have changed the testing to:

``` bash
if [[ "${TXwImage}" == "NONE"* ]] ; then
```

which works well if `T2wInputImages` is set to `NONE`.

@mharms does that satisy your reservation?

Reply:
It sounds like it would.

Reply:
:-) Ok!

Reply:
I don't really follow this, but having any variable contain "NONE" along with other junk (which is what the `*` globbing looks like it is intended to deal with) sounds bad to me.  Also, why is the `*` after the NONE?

Reply:
@coalsont 
To not change the input to the scripts, when T2w is set to `NONE` (as in no image), the `TXwImage` in this script starts with NONE, and that's what the code tests for. If it starts with NONE it means there is no T2w image. The code though now misses setting the `T2wFolder` to NONE if `T2wInputImages` is NONE and that needs to be added to the main script. I'll add that tomorrow.

Reply:
This is now resolved and The T2wPath is set to NONE if T2wImages are set to NONE. It needs to be set to NONE, to avoid adding additional changes to the script, which later requires testing for `"NONE"*` when the path is extended in function calls.

Reply:
I still don't like the testing for `"NONE"*`, it looks like a kludge to me, and @mharms comments of "symmetry of the call" and "variable name contains the word 'path'" don't seem nearly as bad to me (and the variable name could just be changed to use 'full' instead.

@mharms, I think if we want to indicate that a file doesn't exist, it should always either be "" (empty string), or "NONE", and never "NONE/NONE_acpc_dc" or any other garbage.  Please explain why making the code correctly pass "NONE" or "" is worse in this instance.

Reply:
Also, @sivcek, sorry for not disagreeing more strongly earlier.

Reply:
:-) It just feels like I am an yo-yo going back and forth between different options. Doing coding one way and then recoding it all almost back to previous. I'll wait for you guys to reach a consensus and then change accordingly.

Reply:
Sorry for the back-n-forth on this one.

Is there a way to embed a conditional at the point of generating the calls to `T2wToT1wDistortionCorrectAndReg.sh` and `T2wToT1wReg.sh`?
e.g., something like:
```
${RUN} ${HCPPIPEDIR_PreFS}/T2wToT1wDistortionCorrectAndReg.sh \
        --workingdir=${wdir} \
        --t1=${T1wFolder}/${T1wImage}_acpc \
        --t1brain=${T1wFolder}/${T1wImage}_acpc_brain \
    CONDITIONAL HERE on what to use for --t2 and --t2brain
        --fmapmag=${MagnitudeInputName} \
        --fmapphase=${PhaseInputName} \
...
```
I imagine that there is, but that it might be ugly.

As an alternative, my suggestion would be that we construct an explicit argList (a construction that we use in other scripts to good effect, IMO):
```
argList="--workingdir=${wdir} "
argList+="--t1=${T1wFolder}/${T1wImage}_acpc "
```
... etc, with a conditional on what to add for the --t2* options, 
and then invoke with:
`${RUN} ${HCPPIPEDIR_PreFS}/T2wToT1wDistortionCorrectAndReg.sh ${argList}`

If @coalsont doesn't like that approach, then I guess we have indeed come full circle and we revert to @sivcek's original implementation, but with a better variable name for the composite path/file variable than `T2wImagePath` (e.g., `T2wImageWithPath`)

Reply:
Using a variable for the argument seemed perfectly reasonable to me.  I don't like the idea of an inline conditional, those are always terrible to read, and the syntax is obscure (because they are avoided for being ugly and hard to read).  I don't know that an argList concept would be an improvement, and the way you suggested loses any opportunity for whitespace safety.  The safe version (without a conditional for checking NONE) is an array:

```
argList=()
argList+=(--workingdir="$wdir")
argList+=(--t1="$T1wFolder/${T1wImage}_acpc")
...
<script> "${argList[@]}"
```

Personally, I would choose what he originally had, it was the obvious solution to avoid using a garbage "NONE" value inside the construction of a file path.

Reply:
For reference, it looks like bash may not have a purpose-built ternary operator, so an in-place conditional would have to be constructed from shell substitution and either if/else or short circuit operators, like so:

```
    --t1=$( ([[ "$T1wImage" == "NONE" ]] && echo "NONE") || echo "${T1wFolder}/${T1wImage}_acpc")
```

I still don't like the ternary idea.  I endorse using a variable, `T2wImageWithPath` or `T2wImageFull` seem like reasonable variable names.

Reply:
That's actually reasonably "elegant" in-line solution.  I think the argList approach maintains easy readability and comprehension, but admittedly the only real improvement it offers here is relatively minor -- it eliminates the need to search the script for how `T2wImageWithPath` is defined.  But searching for how a variable is defined is something we have to do a lot anyway, so I'm fine with @sivcek reverting to the original implementation with the revised variable name.

Reply:
I hope I have resolved the discussion in an acceptable manner by leaving all the calls as they were written and just adding a `T2wAvailable` TRUE/FALSE variable. I know it is a third option, but it avoids other checks and explicitly states whether T2w image is available or not.

If that is not ok, I can set up:
* `T2wImageWithPath_acpc`
* `T2wImageWithPath_acpc_dc`
variables to use instead of inline path construction and set them to NONE when T2w is not available. So @mharms @coalsont, please a vote on this.

Reply:
I'll accede to whatever @coalsont wants on this one.  The description of what you implemented sounds reasonable to me.

Reply:
It feels weird to pass in garbage filenames that don't exist and use a different argument to say to ignore them, that definitely isn't the convention we use elsewhere.  However, it is not as bad as testing for "NONE"*.  Since these are only internal script calls, and not exposed to the user, I guess that is okay, though I still prefer actually passing a clean "NONE".

Reply:
@coalsont To have clean passing, would this be ok:

```
if [ "${T2wInputImages}" = "NONE" ] ; then
  T2wFolder_T2wImageWithPath_acpc="NONE"
  T2wFolder_T2wImageWithPath_acpc_brain="NONE"
  T1wFolder_T2wImageWithPath_acpc_dc="NONE"
else
  T2wFolder_T2wImageWithPath_acpc="${T2wFolder}/${T2wImage}_acpc"
  T2wFolder_T2wImageWithPath_acpc_brain="${T2wFolder}/${T2wImage}_acpc_brain"
  T1wFolder_T2wImageWithPath_acpc_dc=${T1wFolder}/${T2wImage}_acpc_dc
fi
```

And then calls with these variables and checking within scripts w/o use of `T2wAvailable`?

Reply:
That does look like the solution I prefer, yes.

Reply:
Implemented.